### ISSUE TITLE
MS SQL Server does not have OutParameters and cursors so would need ResultBearing support on Call

### ISSUE THREAD

**Issue by hgschmie** [jdbi/jdbi#2572]
### Discussed in https://github.com/jdbi/jdbi/discussions/2557

<div type='discussions-op-text'>

<sup>Originally posted by **metaforte** December  8, 2023</sup>
Hi,

I have a few MS-SQL server stored procedures that return result set. I am not having any Out Params in the signature.
Is it possible to let

Call.stream<T>() like in Query class

Thanks</div>

**Comment by hgschmie** [comment]
Turns out it is not that simple. MS SQL (and it seems some other databases) support *both* returning results through the regular ResultSet available as `getResultSet()` on the PreparedCall class *and* additional outparameter methods. And then there is this comment on https://learn.microsoft.com/en-us/sql/connect/jdbc/using-a-stored-procedure-with-output-parameters?view=sql-server-ver16:

```
The Microsoft JDBC Driver for SQL Server follows the JDBC 3.0 specification, which states that multiple result sets and update counts should be retrieved before the OUT parameters are retrieved. That is, the application should retrieve all of the ResultSet objects and update counts before retrieving the OUT parameters by using the CallableStatement.getter methods. Otherwise, the ResultSet objects and update counts that haven't already been retrieved will be lost when the OUT parameters are retrieved.
```

What that means is that we can not use the current "fetch all the out parameters right away" code in Call because it would clobber over the ResultSet. And we can not retrieve the result set because the logic in the iterator and ResultBearing will close the context (and the ResultSet and the PreparedCall statement), which in turn makes lazy retrieving the out parameters *after* the result set currently impossible. So *either ResultSet or Outparameters works but not both.*

This is bad. We have to unravel the "close everything on result set iteration end" logic that is a cornerstone of resource management for Jdbi in the case of Call (which is doable but it will turn OutParameters into a closeable object that *must* be managed and all resources related to the outparameter (which is the statement and the resultset need to be attached there and closed when OutParameters goes out of scope.

This will most likely turn into two larger PRs: One that will make the functionality possible but keep the behavior of the current `Call` because it is too convenient for almost every database besides MS SQL (in fact, every database *could* support this but I have not found any other database that does) and a separate `AdvancedCall` (name is work in progress) call for this use case that will require users to manage resources more carefully than with the current `Call` API.

What a mess. I want to like what MS did there with sqlserver, running on linux and all that, but they make it really hard.

**Comment by hgschmie** [comment]
shipped in 3.43.0
